scripts/20_save_data.sh

```diff
@@ -0,0 +1,164 @@
+#!/bin/sh
+#
+# Copyright (C) 2018 Ultimaker B.V.
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
+#
+# SPDX-License-Identifier: AGPL-3.0+
+
+set -eu
+
+trap cleanup EXIT
+
+# system_update wide configuration settings
+SYSTEM_UPDATE_CONF_DIR="${SYSTEM_UPDATE_CONF_DIR:-}"
+TARGET_STORAGE_DEVICE="${TARGET_STORAGE_DEVICE:-}"
+UPDATE_ROOTFS_SOURCE="${UPDATE_ROOTFS_SOURCE:-}"
+UPDATE_SAVE_DIR="${UPDATE_SAVE_DIR:-}"
+# end system_update wide configuration settings
+
+KEEP_LIST="$(mktemp)"
+
+NAME_TEMPLATE_TARGET_MOUNT="um-save_mount"
+TARGET_MOUNT="$(mktemp -d -t "${NAME_TEMPLATE_TARGET_MOUNT}.XXXXXX")"
+
+usage()
+{
+    echo "Usage: ${0} [OPTIONS]"
+    echo "Save the data files from the TARGET_STORAGE_DEVICE."
+    echo "  -d <TARGET_STORAGE_DEVICE>, the target storage device"
+    echo "  -h Print this help text and exit"
+    echo "  -k <SYSTEM_UPDATE_CONF_DIR>, directory that contains files ending with .keep extension containing files and directories to be saved up"
+    echo "  -s <UPDATE_SAVE_DIR>, location that will hold the created archives"
+    echo "Note: that all arguments can also be passed by adding them to the environment."
+}
+
+cleanup()
+{
+    if [ -f "${KEEP_LIST}" ]; then
+        unlink "${KEEP_LIST}"
+    fi
+
+    # On slow media, umount and/or rmdir can fail with 'resource busy' errors.
+    # To do our best with cleanup, attempt this a few times before giving up.
+    failed=0
+    while test -d "${TARGET_MOUNT}"; do
+        if grep -q "${TARGET_MOUNT}" "/proc/mounts"; then
+            if ! umount "${TARGET_MOUNT}"; then
+                failed="$((failed + 1))"
+            fi
+        fi
+
+        if [ -d "${TARGET_MOUNT}" ] && \
+           [ -z "${TARGET_MOUNT##*${NAME_TEMPLATE_TARGET_MOUNT}*}" ]; then
+            if ! rmdir "${TARGET_MOUNT}"; then
+                failed="$((failed + 1))"
+            fi
+        fi
+
+        if [ "${failed}" -ge 300 ]; then
+            echo "Failed to properly cleanup."
+            exit 1
+        fi
+
+        sleep 1
+    done
+}
+
+save_data()
+{
+    echo "Saving data ..."
+
+    for keep in "${SYSTEM_UPDATE_CONF_DIR}/"*".keep" \
+                "${UPDATE_ROOTFS_SOURCE}/${SYSTEM_UPDATE_CONF_DIR}/"*".keep"; do
+        if [ ! -r "${keep}" ]; then
+            continue
+        fi
+
+        # Tar likes its exclude list to be relative, so replace the initial
+        # '/' on a line with './'.
+        sed s,^/,./, "${keep}" >> "${KEEP_LIST}"
+    done
+
+    for partition in "${TARGET_STORAGE_DEVICE}"*; do
+        if [ ! -b "${partition}" ] || \
```

Comment:
Remove || and be verbose

Reply:
You hate 2> /dev/null

Reply:
you are right, I don't want to use 2> /dev/null anymore.
Here I thought it may not be bad, as we may get quite a few errors from it (any partition that cannot be mounted will produce an error, even if that's okay).

As for the removing the or, I'm not sure I understand why